examples/custom_image_embedding_search.ipynb

```diff
@@ -0,0 +1,484 @@
+      "source": [
+        "### Installations and imports\n",
+        "First let's install the relevant packages."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "7hgrcVEl0Ma1"
+      },
+      "outputs": [],
+      "source": [
+        "#installations\n",
+        "!pip install clip\n",
+        "!pip install torch\n",
+        "!pip install pillow\n",
+        "!pip install faiss-cpu\n",
+        "!pip install numpy\n",
+        "!pip install git+https://github.com/openai/CLIP.git\n",
+        "!pip install openai"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "GgrlBLTpT0si"
+      },
+      "source": [
+        "Then let's import all the needed packages.\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 44,
+      "metadata": {
+        "id": "pN1cWF-iyLUg"
+      },
+      "outputs": [],
+      "source": [
+        "#imports\n",
+        "import json\n",
+        "from PIL import Image\n",
+        "from tqdm import tqdm\n",
+        "import torch\n",
+        "from openai import OpenAI\n",
+        "import torch.nn as nn\n",
+        "from torch.utils.data import DataLoader\n",
+        "import clip\n",
+        "import json\n",
+        "import matplotlib.pyplot as plt\n",
+        "import os\n",
+        "import base64\n",
+        "import numpy as np\n",
+        "import pickle\n",
+        "from typing import List, Union, Tuple\n",
+        "import faiss"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "9fONcWxRqll8"
+      },
+      "source": [
+        "Now let's load the CLIP model."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 45,
+      "metadata": {
+        "id": "_Ua9y98NRk70"
+      },
+      "outputs": [],
+      "source": [
+        "#load model on device\n",
+        "device = \"cpu\"\n",
+        "model, preprocess = clip.load(\"ViT-B/32\",device=device)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "wVBAMyhesAyi"
+      },
+      "source": [
+        "Next we will create our image embeddings knowledge base from a directory of images. This will be the knowledge base of technology that we search through to provide information to the user for an image they upload.\n",
+        "\n",
+        "We pass in the directory in which we store our images (as JPEGs) and loop through each to create our embeddings.\n",
+        "\n",
+        "We also have a description.json. This has an entry for every single image in our knowledge base. It has two keys: 'image_path' and 'description'. It maps each image to a useful description of this image to aid in answering the user question."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 46,
+      "metadata": {
+        "id": "vE9i3zLuRk5c"
+      },
+      "outputs": [],
+      "source": [
```

Comment:
i'd split this up into separate code blocks, and put your longer comment explanations as text before. then you can give the tricky bits more space (and comments) e.g. get_features_from_image_path